File: src/enforcer.py
"""
UseTruth verdict enforcer, Python reference SDK (self-contained).

Deny by default: a message is RELEASED only if it carries a fresh, signed ALLOW
attestation bound to that exact claim. No token, ignored BLOCK, replay, tamper,
or a different/appended message all fail closed. Verifies OFFLINE against
UseTruth's published JWKS (https://api.usetruth.ai/.well-known/jwks.json), no
per-message callback.

Spec: SPEC.md. This file is SHIPPABLE as-is (single module).
The canonicalization is vendored verbatim from src/claim_canonical.py; the test
`test_sdk_enforcer_sync` guarantees the copy never drifts.

Dependency: `cryptography` (ES256 verification). Stdlib otherwise.
"""
from __future__ import annotations

import base64
import hashlib
import json
import unicodedata
import urllib.request

RELEASE = "RELEASE"
DENY = "DENY"
JWKS_URL = "https://api.usetruth.ai/.well-known/jwks.json"
_P256 = 32


class CanonicalError(ValueError):
    pass


# --- canonical claim hash (VENDORED from src/claim_canonical.py, keep in sync) ---

def _normalize_content(text) -> str:
    if not isinstance(text, str):
        raise CanonicalError("content must be a string")
    s = unicodedata.normalize("NFC", text).replace("\r\n", "\n").replace("\r", "\n")
    if not s.strip():
        raise CanonicalError("content is empty")
    return s


def _jcs(value) -> str:
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        raise CanonicalError("float is not allowed in a claim hash")
    if isinstance(value, str):
        return json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    if isinstance(value, list):
        return "[" + ",".join(_jcs(v) for v in value) + "]"
    if isinstance(value, dict):
        items = sorted(value.items(), key=lambda kv: str(kv[0]))
        return "{" + ",".join(_jcs(str(k)) + ":" + _jcs(v) for k, v in items) + "}"
    raise CanonicalError(f"unsupported type in claim: {type(value).__name__}")


def claim_hash(content, category, assertions=None, scopes=None) -> str:
    norm_assertions = sorted(
        ({"key": str(a["key"]), "value": a.get("value")}
         for a in (assertions or []) if isinstance(a, dict) and "key" in a),
        key=lambda a: a["key"],
    )
    obj = {
        "content": _normalize_content(content),
        "category": str(category or ""),
        "assertions": norm_assertions,
        "scopes": sorted(str(s) for s in (scopes or [])),
    }
    return hashlib.sha256(_jcs(obj).encode("utf-8")).hexdigest()


# --- JOSE helpers -----------------------------------------------------------

def _b64url_decode(s: str) -> bytes:
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


def _raw_to_der(raw: bytes, size: int = _P256) -> bytes:
    if len(raw) != 2 * size:
        raise ValueError("bad raw signature length")

    def enc(x):
        b = x.to_bytes((x.bit_length() + 7) // 8 or 1, "big")
        if b[0] & 0x80:
            b = b"\x00" + b
        return b"\x02" + bytes([len(b)]) + b

    body = enc(int.from_bytes(raw[:size], "big")) + enc(int.from_bytes(raw[size:], "big"))
    return b"\x30" + bytes([len(body)]) + body


def public_key_from_jwk(jwk):
    from cryptography.hazmat.primitives.asymmetric import ec
    x = int.from_bytes(_b64url_decode(jwk["x"]), "big")
    y = int.from_bytes(_b64url_decode(jwk["y"]), "big")
    return ec.EllipticCurvePublicNumbers(x, y, ec.SECP256R1()).public_key()


def fetch_jwks(url: str = JWKS_URL, timeout: int = 10) -> dict:
    """Fetch and parse UseTruth's JWKS into {kid: public_key}. Cache the result
    in your app (keys rotate rarely) rather than calling per message."""
    with urllib.request.urlopen(url, timeout=timeout) as resp:
        keys = json.loads(resp.read().decode("utf-8")).get("keys", [])
    return {k["kid"]: public_key_from_jwk(k) for k in keys if k.get("kty") == "EC"}


# --- the enforcement decision ----------------------------------------------

def authorize(*, jws, outgoing_text, action, jwks, now, seen_jti, max_skew=60):
    """Deny-by-default authorization. Returns (RELEASE|DENY, reason).

      jws          - the compact JWS attestation the agent presents
      outgoing_text- the exact text about to be sent (rebound against the hash)
      action       - the outgoing action (assertions / claim_scopes for the rebind)
      jwks         - {kid: public_key} (see fetch_jwks)
      now          - current epoch seconds (int)
      seen_jti     - a mutable set used as the replay cache
    """
    parts = str(jws or "").split(".")
    if len(parts) != 3:
        return (DENY, "malformed_token")
    header_b64, payload_b64, sig_b64 = parts
    try:
        header = json.loads(_b64url_decode(header_b64))
        payload = json.loads(_b64url_decode(payload_b64))
    except Exception:  # noqa: BLE001
        return (DENY, "unparseable_token")

    if header.get("alg") != "ES256":
        return (DENY, "bad_alg")
    pub = jwks.get(header.get("kid"))
    if pub is None:
        return (DENY, "unknown_kid")
    from cryptography.hazmat.primitives import hashes
    from cryptography.hazmat.primitives.asymmetric import ec
    try:
        pub.verify(_raw_to_der(_b64url_decode(sig_b64)),
                   (header_b64 + "." + payload_b64).encode("ascii"),
                   ec.ECDSA(hashes.SHA256()))
    except Exception:  # noqa: BLE001
        return (DENY, "bad_signature")

    iat, exp = payload.get("iat"), payload.get("exp")
    if not isinstance(iat, int) or not isinstance(exp, int):
        return (DENY, "bad_times")
    if now + max_skew < iat:
        return (DENY, "not_yet_valid")
    if now >= exp:
        return (DENY, "expired")

    jti = payload.get("jti")
    if not jti or jti in seen_jti:
        return (DENY, "replay_or_missing_jti")

    if payload.get("verdict") != "ALLOW":
        return (DENY, "verdict_" + str(payload.get("verdict")))

    claim = payload.get("claim") or {}
    try:
        recomputed = claim_hash(outgoing_text, claim.get("category"),
                                (action or {}).get("assertions"), (action or {}).get("claim_scopes"))
    except CanonicalError:
        return (DENY, "uncanonicalizable_message")
    if recomputed != claim.get("hash"):
        return (DENY, "message_mismatch")

    seen_jti.add(jti)
    return (RELEASE, "ok")

File: src/enforcer.mjs
// UseTruth verdict enforcer, JavaScript reference SDK (zero dependency, Node >=18).
//
// Deny by default: a message is RELEASED only if it carries a fresh, signed ALLOW
// attestation bound to that exact claim. No token, ignored BLOCK, replay, tamper,
// or a different/appended message all fail closed. Verifies OFFLINE against
// UseTruth's published JWKS (no per-message callback).
//
// Spec: SPEC.md. The canonicalization below is byte-for-byte
// identical to the Python signer (NFC + RFC 8785 JCS + SHA-256), that is the
// whole point of choosing JCS, and the live test proves it cross-language.
import { createHash, createPublicKey, verify as cryptoVerify } from 'node:crypto';

export const RELEASE = 'RELEASE';
export const DENY = 'DENY';
export const JWKS_URL = 'https://api.usetruth.ai/.well-known/jwks.json';

const b64url = (s) => Buffer.from(s, 'base64url');

// --- canonical claim hash (must match src/claim_canonical.py) ----------------

function normalizeContent(text) {
  if (typeof text !== 'string') throw new Error('content must be a string');
  const s = text.normalize('NFC').replace(/\r\n/g, '\n').replace(/\r/g, '\n');
  if (s.trim() === '') throw new Error('content is empty');
  return s;
}

function jcs(v) {
  if (v === null) return 'null';
  if (v === true) return 'true';
  if (v === false) return 'false';
  if (typeof v === 'number') {
    if (!Number.isInteger(v)) throw new Error('float is not allowed in a claim hash');
    return String(v);
  }
  if (typeof v === 'string') return JSON.stringify(v);
  if (Array.isArray(v)) return '[' + v.map(jcs).join(',') + ']';
  if (typeof v === 'object') {
    return '{' + Object.keys(v).sort().map((k) => jcs(String(k)) + ':' + jcs(v[k])).join(',') + '}';
  }
  throw new Error('unsupported type in claim');
}

export function claimHash(content, category, assertions, scopes) {
  const norm = (assertions || [])
    .filter((a) => a && typeof a === 'object' && 'key' in a)
    .map((a) => ({ key: String(a.key), value: a.value }))
    .sort((x, y) => (x.key < y.key ? -1 : x.key > y.key ? 1 : 0));
  const obj = {
    content: normalizeContent(content),
    category: String(category || ''),
    assertions: norm,
    scopes: (scopes || []).map(String).sort(),
  };
  return createHash('sha256').update(jcs(obj), 'utf8').digest('hex');
}

// --- JWKS + verification -----------------------------------------------------

export async function fetchJwks(url = JWKS_URL) {
  const { keys } = await (await fetch(url)).json();
  const out = {};
  for (const k of keys || []) {
    if (k.kty === 'EC') out[k.kid] = createPublicKey({ key: k, format: 'jwk' });
  }
  return out;
}

function verifyEs256(pubKey, signingInput, rawSig) {
  // dsaEncoding 'ieee-p1363' == raw R||S, i.e. the JOSE signature form.
  return cryptoVerify('sha256', Buffer.from(signingInput, 'ascii'),
    { key: pubKey, dsaEncoding: 'ieee-p1363' }, rawSig);
}

// --- the enforcement decision ------------------------------------------------

export function authorize({ jws, outgoingText, action, jwks, now, seenJti, maxSkew = 60 }) {
  const parts = String(jws || '').split('.');
  if (parts.length !== 3) return [DENY, 'malformed_token'];
  const [h64, p64, s64] = parts;
  let header, payload;
  try {
    header = JSON.parse(b64url(h64).toString('utf8'));
    payload = JSON.parse(b64url(p64).toString('utf8'));
  } catch { return [DENY, 'unparseable_token']; }

  if (header.alg !== 'ES256') return [DENY, 'bad_alg'];
  const pub = jwks[header.kid];
  if (!pub) return [DENY, 'unknown_kid'];
  let sigOk = false;
  try { sigOk = verifyEs256(pub, h64 + '.' + p64, b64url(s64)); } catch { sigOk = false; }
  if (!sigOk) return [DENY, 'bad_signature'];

  const { iat, exp } = payload;
  if (!Number.isInteger(iat) || !Number.isInteger(exp)) return [DENY, 'bad_times'];
  if (now + maxSkew < iat) return [DENY, 'not_yet_valid'];
  if (now >= exp) return [DENY, 'expired'];

  const jti = payload.jti;
  if (!jti || seenJti.has(jti)) return [DENY, 'replay_or_missing_jti'];

  if (payload.verdict !== 'ALLOW') return [DENY, 'verdict_' + payload.verdict];

  const claim = payload.claim || {};
  let recomputed;
  try {
    recomputed = claimHash(outgoingText, claim.category, action?.assertions, action?.claim_scopes);
  } catch { return [DENY, 'uncanonicalizable_message']; }
  if (recomputed !== claim.hash) return [DENY, 'message_mismatch'];

  seenJti.add(jti);
  return [RELEASE, 'ok'];
}
